From 7b4ea31990c1c43ad8ea86d42c1e451c85933d87 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.com>
Date: Wed, 15 Jan 2020 14:02:43 +0000
Subject: [PATCH] driver: char: rpivid: Add error handling to the
 legacy device load

The return value from device_create for the legacy device was never
checked or handled. Add the required error handling.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
---
 drivers/char/broadcom/rpivid-mem.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

--- a/drivers/char/broadcom/rpivid-mem.c
+++ b/drivers/char/broadcom/rpivid-mem.c
@@ -201,9 +201,14 @@ static int rpivid_mem_probe(struct platf
 		oldname[3] = 'g';
 		oldname[4] = 'o';
 		oldname[5] = 'n';
-		(void)device_create(priv->class, NULL, priv->devid + 1, NULL,
-				       oldname + 1);
+		dev = device_create(priv->class, NULL, priv->devid + 1, NULL,
+				    oldname + 1);
 		kfree(oldname);
+
+		if (IS_ERR(dev)) {
+			err = PTR_ERR(dev);
+			goto failed_legacy_device_create;
+		}
 	}
 
 	dev_info(priv->dev, "%s initialised: Registers at 0x%08lx length 0x%08lx",
@@ -211,6 +216,8 @@ static int rpivid_mem_probe(struct platf
 
 	return 0;
 
+failed_legacy_device_create:
+	device_destroy(priv->class, priv->devid);
 failed_device_create:
 	class_destroy(priv->class);
 failed_class_create:
